worlds/witness/__init__.py

```diff
@@ -164,13 +157,76 @@ def pre_fill(self):
         if good_items_in_the_game:
             random_good_item = self.multiworld.random.choice(good_items_in_the_game)
 
-            first_check = self.multiworld.get_location(
-                "Tutorial Gate Open", self.player
-            )
+            print(random_good_item)
+
             item = self.items_by_name[random_good_item]
 
-            first_check.place_locked_item(item)
-            self.multiworld.itempool.remove(item)
+            if get_option_value(self.multiworld, self.player, "puzzle_randomization") == 1:
+                self.multiworld.local_early_items[self.player][random_good_item] = 1
+            else:
+                first_check = self.multiworld.get_location(
+                    "Tutorial Gate Open", self.player
+                )
+
+                first_check.place_locked_item(item)
+                self.multiworld.itempool.remove(item)
+
+        itempool_difference = len(self.multiworld.get_unfilled_locations(self.player)) - len(self.multiworld.itempool)
+
+        # fill rest of item pool with junk if there is room
+        if itempool_difference > 0:
+            for i in range(0, itempool_difference):
+                self.multiworld.itempool.append(self.create_item(self.get_filler_item_name()))
+
+        # remove junk, Functioning Brain, useful items (non-door), useful door items in that order until there is room
+        if itempool_difference < 0:
+            junk = [
+                item for item in self.multiworld.itempool
+                if item.classification in {ItemClassification.filler, ItemClassification.trap}
+                and item.name != "Functioning Brain"
+            ]
+
+            f_brain = [item for item in self.multiworld.itempool if item.name == "Functioning Brain"]
+
+            usefuls = [
+                item for item in self.multiworld.itempool
+                if item.classification == ItemClassification.useful
+                and item.name not in StaticWitnessLogic.ALL_DOOR_ITEMS_AS_DICT
+            ]
+
+            removable_doors = [
+                item for item in self.multiworld.itempool
+                if item.classification == ItemClassification.useful
+                and item.name in StaticWitnessLogic.ALL_DOOR_ITEMS_AS_DICT
+            ]
+
+            self.multiworld.per_slot_randoms[self.player].shuffle(junk)
+            self.multiworld.per_slot_randoms[self.player].shuffle(usefuls)
+            self.multiworld.per_slot_randoms[self.player].shuffle(removable_doors)
+
+            removed_junk = False
+            removed_usefuls = False
+            removed_doors = False
+
+            for i in range(itempool_difference, 0):
+                if junk:
+                    self.multiworld.itempool.remove(junk.pop())
```

Comment:
It would be better if you could add the items to a "private" list, remove from there and then only add the final list to the multiworld. Iterating through multiworld.itempool may be slow

Reply:
I feel like I did this with some point to it, like "I want this to react to changes to the itempool that happened through plando". I'll have another look

Reply:
Yes, there's even a comment about it further up. Unless you're referring to only this specific line of code / I'm not fully understanding

Reply:
Wait I need to recheck this code anyway this seems not correct

Reply:
Ok I made it functional, but idk if my changes address your concerns. If not, I might need a bit stronger of a pointer (and the "react to plando" point to be addressed)

Reply:
Oh, hm, i thought plando should work automatically given that useful and advancement should be filled first. I don't remember the exact discussions around plando with from_pool=False, so I'll just ping @alwaysintreble here, maybe he some input on this matter.

The issue I am referring to though is that `self.multiworld.itempool.remove` is slow if the itempool is huge. But if you remove ~0 on average it's not really an issue.

Reply:
That's fair. I did wonder, actually, if there is a way to quickly remove a list of items, at least. But all the options I came across rely on removing all duplicates of sth. I'd want to be able to do

[1, 1, 1, 2, 2, 3].removeSeveral([1, 2, 3]) 
-> [1, 1, 2]

Which if I got to do program my own data structures, should be able to be done in O(n). But I don't think the Python ones really support this...

Reply:
Oh, and also: Yes, this only goes into effect under very specific settings combinations, and then it removes a maximum of 30 items.